[WIP] vSphere Tutorials (#1894)
heptio/velero: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md

---

> title: Velero v1.1 backing up and restoring apps on Pivotal Container Service (PKS)

Reviewer: ```suggestion
title: Backup and Restore Apps on Pivotal Container Service (PKS) using  Velero v1.1
```

author_name: Cormac Hogan
# author_avatar: https://placehold.it/64x64
categories: ['kubernetes']
# Tag should match author to drive author pages
tags: ['Velero']
---
Velero version 1.1 provides support to backup Kubernetes applications deployed on top of PKS, the Pivotal Container Service for vSphere. This post will provide detailed information on how to deploy and configure Velero to backup and restore a stateless application (`nginx`) that is running in a Kubernetes cluster deployed by PKS on vSphere. At this time there is no vSphere plugin for snapshotting stateful applications during a Velero backup. In this case, we rely on a third party program called `restic`. However this post does not include an example of how to backup a stateful application. However this post does not include an example of how to backup a stateful application. That is available in another tutorial which can be found [here](./2019-20-09-Velero-v1-1-Stateful-Backup-vSphere.md).

## Overview of steps

* Download and extract Velero v1.1

> * Deploy and Configure Minio Object store

Reviewer: ```suggestion
* Deploy and Configure a Minio Object store
```

Author: Fixed

* Install Velero using the `velero install` command, ensuring that both `restic` support and a Minio `publicUrl` are included
* Implement steps to support PKS
* Run a test backup/restore of a stateless application that has been deployed on Kubernetes that has itself been provisioned by PKS, the Pivotal Container Service

## What this post does not show

* A demonstration on how to do backup/restore of a stateful application (i.e. PVs)

> * The assumption is that the Kubernetes nodes in your cluster have internet access. This guide does not show how to use a local repository for Velero images

Reviewer: ```suggestion
* The assumption is that the Kubernetes nodes in your cluster have internet access in order to pull the Velero images. This guide does not show how to add images using a local repository
```

Author: Fixed

## Download and extract Velero v1.1

> The [Velero v1.1 image can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.

Reviewer: ```suggestion
The [Velero v1.1 binary can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.
```

## Deploy and Configure Minio Object Store as a backup destination

Velero sends metadata about the objects being backed up to an S3 Object Store. If you do not have an S3 Object Store available, Velero provides the manifest file to create a Minio S3 Object Store on your Kubernetes cluster. This means that all Velero backsup can be kept on-premises.

> * Note: Stateful backups of applications deployed in Kubernetes on vSphere that use the `restic` plugin for backing up Persistent Volumes send the backup data to the same S3 Object Store.

Reviewer: ```suggestion
* Note: Stateful backups of applications deployed in Kubernetes on vSphere that use the `restic` plugin for backing up Persistent Volumes would send the backup data to the same S3 Object Store.
```

This step is to point restic to the correct location of Pods for backup purposes, when K8s is deployed by PKS. First, identify the restic DaemonSet.

```bash
$ kubectl get ds --all-namespaces
NAMESPACE     NAME             DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
kube-system   vrops-cadvisor   3         3         3       3            3           <none>          5d3h
pks-system    fluent-bit       3         3         3       3            3           <none>          5d3h
pks-system    telegraf         3         3         3       3            3           <none>          5d3h
velero        restic           3         3         0       3            0           <none>
```

Next, edit the DaemonSet and change the hostPath setting. The before and after edits are shown below.

```bash
$ kubectl edit ds restic -n velero

# Before
      volumes:
      - hostPath:
          path: /var/lib/kubelet/pods
          type: ""
        name: host-pods

#After
      volumes:
      - hostPath:
          path: /var/vcap/data/kubelet/pods
          type: ""
        name: host-pods

daemonset.extensions/restic edited
```

## Deploy a sample application to backup

> Velero provide a sample `nginx` application for backup testing. This `nginx` deployment assumes the presence of a LoadBalancer for its Service. PKS supports NSX-T integration, so if this is configured, NSX-T will provide this service for you. If you do not have a Load Balancer as part of your Container Network Interface (CNI), there are some easily configuration ones available to get your started. One example is MetalLb, available [here](https://metallb.universe.tf/).

Reviewer: ```suggestion
Velero provides a sample `nginx` application for backup testing. This `nginx` deployment assumes the presence of a LoadBalancer for its Service. PKS supports NSX-T integration, so if this is configured, NSX-T will provide this service for you. If you do not have a Load Balancer as part of your Container Network Interface (CNI), there are some easily configuration ones available to get your started. One example is MetalLb, available [here](https://metallb.universe.tf/).
```

```

```bash
$ velero restore create nginx-restore --from-backup nginx-backup
Restore request "nginx-restore" submitted successfully.
Run `velero restore describe nginx-restore` or `velero restore logs nginx-restore` for more details.
```

The following command can be used to examine the restore in detail, and check to see if it has successfully completed.

```bash
$ velero restore describe nginx-restore
Name:         nginx-restore
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:  Completed

Backup:  nginx-backup

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto
```

## Verify that the restore succeeded

You can see that the restore has now completed. Check to see if the namespace, DaemonSet and service has been restored using the `kubectl` commands shown previously. One item to note is that the `nginx` service may be restored with a new IP address from the LoadBalancer. This is normal.

> Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes I can! Looks like the restore was successful.

Reviewer: ```suggestion
Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes we can! Looks like the restore was successful.
```

![nginx restored](../img/vsphere-tutorial-icons/nginx-restore-new-ip.png)

> Backups and Restore are now working on Kubernetes deployed by PKS on vSphere using Velero v1.1.

Reviewer: ```suggestion
Backups and Restores are now working on Kubernetes deployed by PKS on vSphere using Velero v1.1.
```
